### PR TITLE
HTML API: Add support for list elements.

### PR DESCRIPTION
**Pull request by dmsnell**
Trac ticket: Core-60215

Adds support to the HTML Processor for the following elements, enabling the handling of list elements in HTML:

 - UL, OL, LI
 - DT, DL, DD

Previously, the HTML Processor would bail when encountering any of these tags. Now it will properly handle them and their complicated rules.

- [x] Add comment at top of HTML Processor indicating new tag support.
- [ ] Needs a comprehensive suite of tests due to the complicated nature of list rules.
- [x] Add docblocks to new tests.

### CODE REVIEW

File: src/wp-includes/html-api/class-wp-html-open-elements.php [hunk 1]
```diff
@@ -429,4 +459,15 @@ public function after_element_pop( $item ) {
...
+	/**
+	 * Represents the collection of H1-H6 elements.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @see has_element_in_scope()
+	 *
+	 * @var string
+	 */
+	const HEADING_ELEMENTS = 'heading-elements';
```
Comment (dmsnell) [inline src/wp-includes/html-api/class-wp-html-open-elements.php]: maybe move this to the Tag Processor? then we could use it in `next_tag()` too

File: src/wp-includes/html-api/class-wp-html-processor.php [hunk 2]
```diff
@@ -751,6 +755,98 @@ private function step_in_body() {
...
+				in_body_list_loop:
+				/*
+				 * The logic for LI and DT/DD is the same except for one point: LI elements _only_
+				 * close other LI elements, but a DT or DD element closes _any_ open DT or DD element.
+				 */
+				if ( $is_li ? 'LI' === $node->node_name : ( 'DD' === $node->node_name || 'DT' === $node->node_name ) ) {
+					$node_name = $is_li ? 'LI' : $node->node_name;
+					$this->generate_implied_end_tags( $node_name );
+					if ( $node_name !== $this->state->stack_of_open_elements->current_node()->node_name ) {
+						// @TODO: Indicate a parse error once it's possible. This error does not impact the logic here.
```
Comment (sirreal) [inline src/wp-includes/html-api/class-wp-html-processor.php:778]:
```suggestion
						// @todo Indicate a parse error once it's possible. This error does not impact the logic here.
```

File: src/wp-includes/html-api/class-wp-html-processor.php [hunk 3]
```diff
@@ -751,6 +755,98 @@ private function step_in_body() {
...
+					) ||
+					(
+						'LI' !== $tag_name &&
+						! $this->state->stack_of_open_elements->has_element_in_scope( $tag_name )
+					)
+				) {
+					/*
+					 * This is a parse error, ignore the token.
+					 *
+					 * @TODO: Indicate a parse error once it's possible.
```
Comment (sirreal) [inline src/wp-includes/html-api/class-wp-html-processor.php:836]:
```suggestion
					 * @todo Indicate a parse error once it's possible.
```

File: src/wp-includes/html-api/class-wp-html-processor.php [hunk 4]
```diff
@@ -751,6 +755,98 @@ private function step_in_body() {
...
+					 */
+					return $this->step();
+				}
+
+				$this->generate_implied_end_tags( $tag_name );
+
+				if ( $tag_name !== $this->state->stack_of_open_elements->current_node()->node_name ) {
+					// @TODO: Indicate a parse error once it's possible. This error does not impact the logic here.
```
Comment (sirreal) [inline src/wp-includes/html-api/class-wp-html-processor.php:844]:
```suggestion
					// @todo Indicate a parse error once it's possible. This error does not impact the logic here.
```

**Review by sirreal — COMMENTED** [review]
I ran the suite from #5794 against this PR and this case fails:

```
#data
<ul><li><ul></li><li>a</li></ul></li></ul>
#errors
(1,4): expected-doctype-but-got-start-tag
(1,17): unexpected-end-tag
#document
| <html>
|   <head>
|   <body>
|     <ul>
|       <li>
|         <ul>
|           <li>
|             "a"
```

Here's the failure:

```
 ✘ External html5lib with tests1/case104·-·line·1611
   ┐
   ├ HTML <ul><li><ul></li><li>a</li></ul></li></ul>
   ├  was not processed correctly.
   ├ Failed asserting that two strings are equal.
   ┊ ---·Expected
   ┊ +++·Actual
   ┊ @@ @@
   ┊      <ul>\n
   ┊        <li>\n
   ┊          <ul>\n
   ┊ -··········<li>\n
   ┊ +······<li>\n
   ┊  '
```

**Comment by sirreal** [comment]
I'll see if I can figure out the test failure. My previous comment was wrong (the expect/actual were inverted), I've updated it.

**Comment by sirreal** [comment]
I've rebased this.

I pushed a test and a fix for the issue the html5lib tests found. I believe `has_element_in_specific_scope` had a flipped boolean when it finds a tag in the termination list. Just like if it reaches `HTML`, the search ends without having found a the tag in scope and the search ends in failure.

This fixes the tests.

**Review by sirreal — APPROVED** [review]
I'm happy with this, I think it's ready to ship. We've given it a good amount of testing.

Thanks!

File: src/wp-includes/html-api/class-wp-html-processor.php [hunk 5]
```diff
@@ -755,6 +759,109 @@ private function step_in_body() {
...
+				}
+
+				if (
+					'ADDRESS' !== $node->node_name &&
+					'DIV' !== $node->node_name &&
+					'P' !== $node->node_name &&
+					$this->is_special( $node->node_name )
+				) {
+					/*
+					 * > If node is in the special category, but is not an address, div,
```
Comment (audrasjb) [inline src/wp-includes/html-api/class-wp-html-processor.php:796]: The `>` character should probably be removed here

File: src/wp-includes/html-api/class-wp-html-processor.php [hunk 6]
```diff
@@ -755,6 +759,109 @@ private function step_in_body() {
...
+					 * > or p element, then jump to the step labeled done below.
+					 */
+					goto in_body_list_done;
+				} else {
+					/*
+					 * > Otherwise, set node to the previous entry in the stack of open elements
```
Comment (audrasjb) [inline src/wp-includes/html-api/class-wp-html-processor.php:802]: The `>` character should probably be removed here

File: src/wp-includes/html-api/class-wp-html-processor.php [hunk 7]
```diff
@@ -755,6 +759,109 @@ private function step_in_body() {
...
+				in_body_list_done:
+				if ( $this->state->stack_of_open_elements->has_p_in_button_scope() ) {
+					$this->close_a_p_element();
+				}
+
+				$this->insert_html_element( $this->state->current_token );
+				return true;
+
+			/*
+			 * > An end tag whose tag name is "li"
```
Comment (audrasjb) [inline src/wp-includes/html-api/class-wp-html-processor.php:821]: The `>` character should probably be removed here

File: tests/phpunit/tests/html-api/wpHtmlProcessorSemanticRules.php [hunk 8]
```diff
@@ -224,6 +224,107 @@ public function test_in_body_button_with_button_in_scope_as_ancestor() {
 		$this->assertSame( array( 'HTML', 'BODY', 'BUTTON' ), $p->get_breadcrumbs(), 'Failed to produce expected DOM nesting for third button.' );
 	}
 
+	/**
+	 * Verifies that H1 through H6 elements close an open P element.
+	 *
+	 * @ticket {TICKET_NUMBER}
```
Comment (audrasjb) [inline tests/phpunit/tests/html-api/wpHtmlProcessorSemanticRules.php]: Ticket number is `60215`

**Review by audrasjb — COMMENTED** [review]
Thanks for the PR. I added a few comments for small enhancements.

File: src/wp-includes/html-api/class-wp-html-processor.php (hunk 5 above)
Comment (sirreal) [reply src/wp-includes/html-api/class-wp-html-processor.php:796]: I believe the `>` is included to indicate it's a quote. These are quoting the HTML standard directly.

Comment (dmsnell) [reply src/wp-includes/html-api/class-wp-html-processor.php:796]: Confirming what @sirreal wrote. These are linked to the HTML5 living spec by words to make it easier to search, and the `>` is attempting to indicate that it's a quote. It also corresponds to every block within the `step_in_body()` function.

I'm happy to entertain suggestions to improve the linking, but I don't want to make it more difficult for people to cross-reference the spec by removing these quotes. Unfortunately the spec doesn't add anchors to each of these sections so there's no clear link (which is one of the original reasons I added the quotes, beyond the obvious that it puts the intention of the spec at the front in the code).

**Comment by dmsnell** [comment]
Merged in [[57264]](https://core.trac.wordpress.org/changeset/57264)

File: src/wp-includes/html-api/class-wp-html-processor.php [hunk 9]
```diff
@@ -105,7 +105,7 @@
  *  - Formatting elements: B, BIG, CODE, EM, FONT, I, SMALL, STRIKE, STRONG, TT, U.
  *  - Heading elements: H1, H2, H3, H4, H5, H6, HGROUP.
  *  - Links: A.
- *  - Lists: DL.
+ *  - Lists: DD, DL, DT, LI, OL, LI.
```
Comment (mdawaffe) [inline src/wp-includes/html-api/class-wp-html-processor.php:108]: This looks like it should this be `UL`.

```diff
- *  - Lists: DD, DL, DT, LI, OL, LI.
+ *  - Lists: DD, DL, DT, LI, OL, UL.
```

Comment (dmsnell) [reply src/wp-includes/html-api/class-wp-html-processor.php:108]: Thank you! Will fix this.

**Pull request closed without merging**